### PR TITLE
implement async step; update debugger library

### PR DESCRIPTION
**Pull request by devoncarew**
- implement async stepping (fix https://youtrack.jetbrains.com/issue/WEB-22910)
- update the debugger library to the latest; this properly supports optional fields in the protocol (and avoids an NPE), and exposes a few fields that had not yet been exposed

@alexander-doroshko, @jwren, @danrubel

This implementation essentially store the `atAsyncSuspension` field when an isolate pauses, and when the user steps, we either issue a `StepOption.OverAsyncSuspension` or a `StepOption.Over`.

### CODE REVIEW

**Comment by danrubel** [comment]
LGTM

File: Dart/src/com/jetbrains/lang/dart/ide/runner/server/vmService/DartVmServiceListener.java [hunk 1]
```diff
@@ -129,6 +136,8 @@ private void onIsolatePaused(@NotNull final IsolateRef isolateRef,
       final boolean suspend = myDebugProcess.getSession().breakpointReached(xBreakpoint, logExpression, suspendContext);
       if (!suspend) {
         myDebugProcess.getVmServiceWrapper().resumeIsolate(isolateRef.getId(), null);
+      } else {
+        myDebugProcess.getSession().positionReached(suspendContext);
```
Comment (alexander-doroshko) [inline Dart/src/com/jetbrains/lang/dart/ide/runner/server/vmService/DartVmServiceListener.java]: `breakpointReached()` is already called 3 lines above, why do you think you need to call `positionReached()`?

File: Dart/src/com/jetbrains/lang/dart/ide/runner/server/vmService/DartVmServiceListener.java [hunk 2]
```diff
@@ -118,6 +119,12 @@ private void onIsolatePaused(@NotNull final IsolateRef isolateRef,
 
       final XLineBreakpoint<XBreakpointProperties> xBreakpoint = myBreakpointHandler.getXBreakpoint(vmBreakpoints.get(0));
 
+      if (xBreakpoint == null) {
+        myLatestSourcePosition = sourcePosition;
+        myDebugProcess.getSession().positionReached(suspendContext);
+        return;
+      }
```
Comment (alexander-doroshko) [inline Dart/src/com/jetbrains/lang/dart/ide/runner/server/vmService/DartVmServiceListener.java]: I guess this `if` is for stopping on breakpoints set in the Observatory, right?
If so, please make separate feature as a separate commit next time. Thanks!

**Comment by alexander-doroshko** [comment]
Sorry for delay and thanks for the feature! Pushed.
I've only removed `positionReached()` call after `breakpointReached()`.
Next time please prefer separate commits for separate features. For example: update VM Service drivers; handle breakpoints set in the Observatory; async step.

File: Dart/src/com/jetbrains/lang/dart/ide/runner/server/vmService/DartVmServiceListener.java (hunk 2 above)
Comment (devoncarew) [inline Dart/src/com/jetbrains/lang/dart/ide/runner/server/vmService/DartVmServiceListener.java]: The VM implements the async step feature by setting temporary VM-only breakpoints. When we pause, we see it as stopping at a breakpoint, but the location doesn't match up to any of our known breakopints.

File: Dart/src/com/jetbrains/lang/dart/ide/runner/server/vmService/DartVmServiceListener.java (hunk 1 above)
Comment (devoncarew) [inline Dart/src/com/jetbrains/lang/dart/ide/runner/server/vmService/DartVmServiceListener.java]: Hmm, I suppose I was unaware that breakpointReached() implied positionReached()? In any case, thanks for the  review! I'll re-test once this lands.

**Comment by devoncarew** [comment]
> Next time please prefer separate commits for separate features. For example: update VM Service drivers; handle breakpoints set in the Observatory; async step.

Will do re: separate features per PR; in this particular case, this was dealing w/ the internal implementation of async stepping.

File: Dart/src/com/jetbrains/lang/dart/ide/runner/server/vmService/DartVmServiceListener.java (hunk 2 above)
Comment (alexander-doroshko) [inline Dart/src/com/jetbrains/lang/dart/ide/runner/server/vmService/DartVmServiceListener.java]: Ok, thanks for letting me know! I'll add comment to the code.
